cpp/src/rolling/detail/range_utils.cuh

```diff
@@ -413,209 +413,226 @@ template <typename T, typename V>
   }
 }
 
+template <direction Direction, typename OrderbyT, typename Comparator>
+__device__ inline auto compute_bounded(column_device_view::const_iterator<OrderbyT> const& begin,
+                                       size_type i,
+                                       size_type start,
+                                       size_type end,
+                                       OrderbyT value,
+                                       Comparator&& comp)
+{
+  if constexpr (Direction == direction::PRECEDING) {
+    return 1 +
+           thrust::distance(
+             thrust::lower_bound(thrust::seq, begin + start, begin + end, value, comp), begin + i);
+  } else {
+    return thrust::distance(
+             begin + i, thrust::upper_bound(thrust::seq, begin + start, begin + end, value, comp)) -
+           1;
+  }
+};
+
 /**
- * @brief Functor to dispatch computation of clamped range-based rolling window bounds.
+ * @brief Functor to compute distance from a given row to the edge
+ * of the window.
  *
- * @tparam Direction The direction (preceding or following) of the window
- * @tparam WindowTag The tag indicating the type of window being computed
- * @tparam Order The sort order of the orderby column defining the window.
+ * @tparam Grouping Object defining how the orderby column is
+ * grouped.
+ * @tparam OrderbyT Type of elements in the orderby columns.
+ * @tparam DeltaT Type of the elements in the scalar delta (returned
+ * by scalar.data()).
+ * @param groups The grouping object.
+ * @param row_delta Pointer to row delta on device.
+ * @param begin Iterator to begin of orderby column on device.
+ * @param end Iterator to end of orderby column on device.
+ *
+ * @note Let `x` be the value of the current row and `delta` the provided
+ * row delta then for bounded windows the endpoints are computed as follows.
+ *
+ *           | ASCENDING | DESCENDING
+ * ----------+-----------+-----------
+ * PRECEDING | x - delta | x + delta
+ * FOLLOWING | x + delta | x - delta
+ *
+ * See `saturating_sub` and `saturating_add` for details of the implementation of
+ * saturating addition/subtraction.
  */
-template <direction Direction, window_tag WindowTag, cudf::order Order>
-struct range_window_clamper {
+template <window_tag WindowTag,
+          direction Direction,
+          cudf::order Order,
+          typename Grouping,
+          typename OrderbyT,
+          typename DeltaT>
+struct distance_kernel {
```

Comment:
Removing the kernel from the window functor for benchmarking purposes (next step would be removing the template params).